Sources/PackageGraph/DependencyMirrors.swift

```diff
@@ -0,0 +1,139 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+import TSCBasic
+import TSCUtility
+
+/// A collection of dependency mirrors.
+public final class DependencyMirrors {
+
+    /// A dependency mirror error.
+    public enum Error: Swift.Error {
+        /// No mirror was found for the specified URL.
+        case mirrorNotFound
+    }
+
+    private var storage: [String: String]
+
+    private init(_ mirrors: [Mirror]) {
+        self.storage = Dictionary(mirrors.map({ ($0.original, $0.mirror) }), uniquingKeysWith: { first, _ in first })
+    }
+
+    /// Sets a mirror URL for the given URL.
+    public func set(mirrorURL: String, forURL url: String) {
+        storage[url] = mirrorURL
+    }
+
+    /// Unsets a mirror for the given URL.
+    /// - Parameter originalOrMirrorURL: The original URL or the mirrored URL
+    /// - Throws: `Error.mirrorNotFound` if no mirror exists for the provided URL.
+    public func unset(originalOrMirrorURL: String) throws {
+        if storage.keys.contains(originalOrMirrorURL) {
+            storage[originalOrMirrorURL] = nil
+        } else if let mirror = storage.first(where: { $0.value == originalOrMirrorURL }) {
+            storage[mirror.key] = nil
+        } else {
+            throw Error.mirrorNotFound
+        }
+    }
+
+    /// Returns the mirrored URL for a package dependency URL.
+    /// - Parameter url: The original URL
+    /// - Returns: The mirrored URL, if one exists.
+    public func getMirror(forURL url: String) -> String? {
+        return storage[url]
+    }
+
+    /// Returns the effective URL for a package dependency URL.
+    /// - Parameter url: The original URL
+    /// - Returns: The mirrored URL if it exists, otherwise the original URL.
+    public func effectiveURL(forURL url: String) -> String {
+        return getMirror(forURL: url) ?? url
+    }
+}
+
+extension DependencyMirrors: Collection {
+    public typealias Index = Dictionary<String, String>.Index
+    public typealias Element = String
+
+    public var startIndex: Index {
+        storage.startIndex
+    }
+
+    public var endIndex: Index {
+        storage.endIndex
+    }
+
+    public subscript(index: Index) -> Element {
+        storage[index].value
+    }
+
+    public func index(after index: Index) -> Index {
+        storage.index(after: index)
+    }
+}
+
+extension DependencyMirrors: ExpressibleByDictionaryLiteral {
+    public convenience init(dictionaryLiteral elements: (String, String)...) {
+        self.init(elements.map { Mirror(original: $0.0, mirror: $0.1) })
+    }
+}
+
+extension DependencyMirrors: JSONMappable, JSONSerializable {
```

Comment:
does this need to be using Codable instead?

Reply:
Both here and [here](https://github.com/apple/swift-package-manager/pull/3002#discussion_r515305732), I'm deferring to the original implementation, which uses these pre-Codable JSON serialization protocols. I'd be just as happy to switch to Codable, but I didn't know what our current disposition is (i.e. _stare decisis_ vs. "clean up as we go along").

Reply:
the diff made me think this is new code (where we would prefer Codable), but if its just moving existing code around I think we can focus the PR on what its actually trying to achieve.

Reply:
Agreed; separately we should move everything to Codable, but as separately testable PRs.